pp/serv/rpc.go

```diff
@@ -0,0 +1,157 @@
+package serv
+
+import (
+	"github.com/stratosnet/sds/utils/crypto/secp256k1"
+	"encoding/hex"
+	"strconv"
+	b64 "encoding/base64"
+	"crypto/sha256"
+	"time"
+	"sync"
+	"github.com/stratosnet/sds/pp/file"
+	"github.com/stratosnet/sds/pp/peers"
+	"github.com/stratosnet/sds/pp/requests"
+	"github.com/stratosnet/sds/msg/header"
+	"github.com/stratosnet/sds/utils"
+	"github.com/stratosnet/sds/rpc"
+	rpc_api "github.com/stratosnet/sds/pp/api/rpc"
+)
+
+const (
+	// the length of request shall be shorter than 5242880 bytes
+	// this equals 3932160 bytes after
+	FILE_DATA_SAFE_SIZE = 3500000
+)
+
+
+var (
+	// key: fileHash value: file
+	FileOffset = make(map[string]*FileFetchOffset)
+	FileOffsetMutex sync.Mutex
+)
+
+type FileFetchOffset struct {
+	RemoteRequested    uint64
+	ResourceNodeAsked  uint64
+}
+
+type rpcApi struct {
+
+}
+
+func RpcApi() *rpcApi {
+	return &rpcApi{}
+}
+
+// apis returns the collection of built-in RPC APIs.
+func apis() []rpc.API {
+	return []rpc.API{
+		{
+			Namespace: "user",
+			Version:   "1.0",
+			Service:   RpcApi(),
+			Public:    true,
+		},
+	}
+}
+
+func ResultHook(r []string, fileHash string) []string {
+	if r[0] == rpc_api.UPLOAD_DATA {
+		start, _ := strconv.ParseUint(r[1], 10, 64)
+		end, _ := strconv.ParseUint(r[2], 10, 64)
+		// have to cut the requested data block into smaller pieces when the size is greater than the limit
+		if end - start > FILE_DATA_SAFE_SIZE {
+			f := &FileFetchOffset{RemoteRequested: start + FILE_DATA_SAFE_SIZE, ResourceNodeAsked: end}
+
+			FileOffsetMutex.Lock()
+			FileOffset[fileHash] = f
+			FileOffsetMutex.Unlock()
+
+			var nr []string
+			nr = append(nr, r[0])
+			nr = append(nr, strconv.FormatUint(start, 10))
+			nr = append(nr, strconv.FormatUint(start + FILE_DATA_SAFE_SIZE, 10))
+			return nr
+		}
+	}
+	return r
+}
+
+func (api *rpcApi) RequestUpload(fileName string, fileSize string, fileHash string, walletAddr string, pubkey string, signature string) []string {
+
+	size, err := strconv.Atoi(fileSize)
```

Comment:
why not directly make the file size a number?

Reply:
Because the parameters in json are listed in an array. It would be confusing that some items are strings, some are numbers.
"params": [1, "file.name", 69, "392"]   vs.  "params": ["1", "file.name", "392", "69"] 
I feel the latter sounds better.

Reply:
Now, both params and result have been changed to json objects instead of an array with no names, as demonstrated in the description of this PR.